consensus/hotstuff/cruisectl/block_rate_controller.go

```diff
@@ -0,0 +1,160 @@
+		config:      config,
+		log:         log,
+		state:       state,
+		viewChanges: make(chan uint64, 1),
+		epochSetups: make(chan *flow.Header, 1),
+	}
+
+	ctl.Component = component.NewComponentManagerBuilder().
+		AddWorker(ctl.processEventsWorker).
+		Build()
+
+	// TODO initialize last measurement
+	// TODO initialize epoch info
+	_ = ctl.lastMeasurement
+	_ = ctl.curEpochTargetSwitchover
+	_ = ctl.curEpochFinalView
+	_ = ctl.nextEpochFinalView
+
+	return ctl, nil
+}
+
+// BlockRateDelay returns the current block rate delay value to use when proposing, in milliseconds.
+// This function reflects the most recently computed output of the PID controller
+func (ctl *BlockRateController) BlockRateDelay() float64 {
+	return ctl.blockRateDelay.Load()
+}
+
+// processEventsWorker is a worker routine which processes events received from other components.
+func (ctl *BlockRateController) processEventsWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	done := ctx.Done()
+
+	for {
+		select {
+		case <-done:
+			return
+		case enteredView := <-ctl.viewChanges:
+			err := ctl.processOnViewChange(enteredView)
+			if err != nil {
+				ctl.log.Err(err).Msgf("fatal error handling OnViewChange event")
+				ctx.Throw(err)
+			}
+		case block := <-ctl.epochSetups:
+			snapshot := ctl.state.AtHeight(block.Height)
+			err := ctl.processEpochSetupPhaseStarted(snapshot)
+			if err != nil {
+				ctl.log.Err(err).Msgf("fatal error handling EpochSetupPhaseStarted event")
+				ctx.Throw(err)
+			}
+		}
+	}
+}
+
+// processOnViewChange processes OnViewChange events from HotStuff.
+// Whenever the view changes, we:
+//   - take a new measurement for instantaneous and EWMA block rate
+//   - compute a new target block rate (set-point)
+//   - compute error terms, compensation function output, and new block rate delay
+//   - updates epoch info, if this is the first observed view of a new epoch
+func (ctl *BlockRateController) processOnViewChange(view uint64) error {
+	// TODO
+	return nil
+}
+
+// processEpochSetupPhaseStarted processes EpochSetupPhaseStarted events from the protocol state.
+// Whenever we enter the EpochSetup phase, we:
+//   - store the next epoch's final view
+func (ctl *BlockRateController) processEpochSetupPhaseStarted(snapshot protocol.Snapshot) error {
+	// TODO
+	return nil
+}
+
+// OnViewChange responds to a view-change notification from HotStuff.
+// The event is queued for async processing by the worker. If the channel is full,
+// the event is discarded - since we are taking an average it doesn't matter if
+// occasionally miss a sample.
+func (ctl *BlockRateController) OnViewChange(_, newView uint64) {
+	select {
+	case ctl.viewChanges <- newView:
+	default:
+	}
+}
+
+// EpochSetupPhaseStarted responds to the EpochSetup phase starting for the current epoch.
+// The event is queued for async processing by the worker.
+func (ctl *BlockRateController) EpochSetupPhaseStarted(_ uint64, first *flow.Header) {
```

Comment:
According to documentation this will be hooked up to event distributor and we require that consumers are not-blocking, but this one can block if buffer is full

Reply:
Yes, but we can't miss these events. We'll set the buffer size high enough that getting full is exceedingly unlikely, given that we expect these events very infrequently.